text/3366-diagnostic-attribute-namespace.md

```diff
@@ -0,0 +1,222 @@
+* Change this behaviour at any time
+
+The compiler must:
+
+* Verify that the attribute is syntactically correct, which means:
+    + Its one of the accepted attributes
+    + It only contains the allowed options
+    + Any provided option follows the defined syntax
+* Follow the described semantics if the attribute is not ignored.
+
+
+## The `#[diagnostic::on_unimplemented]` attribute
+
+This section describes the syntax of the `on_unimplemented` attribute and additionally how it is supposed to work. Implementing the later part is optional as outlined in the previous section. Any syntax not explicitly described in this section must be rejected as invalid according to the general rules of the `#[diagnostic]` attribute namespace.
+
+```rust
+#[diagnostic::on_unimplemented(
+    message="message",
+    label="label",
+    note="note",
+)]
+trait MyIterator<A> {
+    fn next(&mut self) -> A;
+}
+```
+
+
+Each of the options `message`, `label` and `note` are optional. They are separated by comma. The tailing comma is optional. Specifying any of these options hints the compiler to replace the normally emitted part of the error message with the provided string. At least one of these options needs to exist. Each option can appear at most once. The error message can include type information for the `Self` type or any generic type by using `{Self}` or `{A}` (where `A` refers to the generic type name in the definition). These placeholders are replaced by the actual type name.
+
+In addition the `on_unimplemented` attribute provides mechanisms to specify for which exact types a certain message should be emitted via an `if()` option. It accepts a set of filter options. A filter option consists on the generic parameter name from the trait definition and a type path against which the parameter should be checked. These type path could either be a fully qualified path or refer to any type in the current scope. As a special generic parameter name `Self` is added to refer to the `Self` type of the trait implementation. A filter option evaluates to `true` if the corresponding generic parameter in the trait definition matches the specified type. The provided `message`/`note`/`label` options are only emitted if the filter operation evaluates to `true`.
+
+The `any` and `all` option allows to combine multiple filter options. The `any` option matches if one of the supplied filter options evaluates to `true`, the `all` option requires that all supplied filter options evaluate to true. `not` allows to negate a given filter option. It evaluates to `true` if the inner filter option evaluates to `false`. These options can be nested to construct complex filters.
+
+The `on_unimplemented` attribute can be applied multiple times to the same trait definition. Multiple attributes are evaluated in order. The first matching instance for each of the `message`/`label`/`note` options is emitted.
```

Comment:
I feel like the bulk of the `on_unimplemented` behavior should be on its own RFC, because there are open questions about it. What happens when you have multiple cases that match and the attribute writer doesn't notice? Is that detected by the compiler and emits an error?

Reply:
Yeah, the discussion of `on_implemented` is so involved that it looks like it's proposed to be stabilized in this RFC. If it isn't, it needs to be trimmed down from irrelevant details and more prominently marked as just an example.

Reply:
I've opted to put everything into one RFC for now as the proposed `#[diagnostic]` attribute namespace is only useful with actual attributes that can be used by someone. So yes: This RFC proposes both, the namespace and the `#[on_unimplemented]` attribute. If it turns out that there is no conses around how the `filter` rules should be designed I would propose to only specify the `#[on_unimplemented]` attribute without the filtering functionality. So just the `message`/`label`/`note` options. I would expect that it is much easier to get these options approved.

Reply:
I think that's a reasonable approach: land `#[diagnostics::on_unimplemented(note="", message="", label="")]` with this RFC and a subsequent RFC for the evolution of `on_unimplemented`. This would also (I think) help t-lang to experience what evolving these would look like in the future and catch any potential problems or concerns *with a concrete example*.

Reply:
I would nevertheless wait for t-lang before removing the relevant RFC section.This would at least give some initial feedback on what needs to be addressed in a future RFC.